Address Safer cpp failures in BitmapImageDescriptor

iidmsa opened the pull request:
#### 0aa86d7e1238be07bce70f2fd692d72468993c7a
<pre>
Address Safer cpp failures in BitmapImageDescriptor
<a href="https://bugs.webkit.org/show_bug.cgi?id=288767">https://bugs.webkit.org/show_bug.cgi?id=288767</a>
<a href="https://rdar.apple.com/problem/145790133">rdar://problem/145790133</a>

Reviewed by Chris Dumez.

Fix a clang static analyzer warning in BitmapImageDescriptor.

* Source/WebCore/platform/graphics/BitmapImageDescriptor.cpp:
(WebCore::BitmapImageDescriptor::BitmapImageDescriptor):
(WebCore::BitmapImageDescriptor::imageMetadata const):
(WebCore::BitmapImageDescriptor::primaryImageFrameMetadata const):
(WebCore::BitmapImageDescriptor::primaryNativeImageMetadata const):
(WebCore::BitmapImageDescriptor::sourceSize const):
(WebCore::BitmapImageDescriptor::singlePixelSolidColor const):
(WebCore::BitmapImageDescriptor::maximumSubsamplingLevel const):
(WebCore::BitmapImageDescriptor::shouldUseQuickLookForFullscreen const):
(WebCore::BitmapImageDescriptor::isSpatial const):
* Source/WebCore/platform/graphics/BitmapImageDescriptor.h:

Canonical link: <a href="https://commits.webkit.org/291436@main">https://commits.webkit.org/291436@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/163a1de77a7964eb55d8bdf8fd031368545c41c8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/92863 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/12414 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/2060 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/97859 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/43389 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/94913 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/12694 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/20866 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/71028 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/28458 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/95865 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/9569 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/84006 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/51356 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/9261 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/1639 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/42702 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/79562 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/1612 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/99881 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/19915 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/14605 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/80047 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/20166 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/79902 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/79347 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19705 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/23887 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/1151 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/12941 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/19899 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/25075 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/19586 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/23046 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/21327 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/platform/graphics/BitmapImageDescriptor.h:
```diff
@@ -123,7 +123,7 @@ class BitmapImageDescriptor {
     mutable SubsamplingLevel m_maximumSubsamplingLevel { SubsamplingLevel::Default };
 
-    BitmapImageSource& m_source;
+    Ref<BitmapImageSource> m_source;
```
This is definitely not OK. This introduces a reference cycle between the `BitmapImageSource` and the `BitmapImageDescriptor`, thus leaking memory. Replacing a `T&` with a `Ref<T>` is dangerous and rarely the right thing to do. You want a `CheckedRef` or a `WeakRef`.

cdumez commented on Source/WebCore/platform/graphics/BitmapImageDescriptor.cpp:
```diff
@@ -32,8 +32,8 @@
 
-BitmapImageDescriptor::BitmapImageDescriptor(BitmapImageSource& source)
-    : m_source(source)
+BitmapImageDescriptor::BitmapImageDescriptor(CheckedRef<BitmapImageSource>&& source)
```
We usually do not pass `CheckedRef` as parameter, this should just be a `BitmapImageSource&`

cdumez commented on Source/WebCore/platform/graphics/BitmapImageDescriptor.cpp:
```diff
@@ -32,8 +32,8 @@
-BitmapImageDescriptor::BitmapImageDescriptor(BitmapImageSource& source)
-    : m_source(source)
+BitmapImageDescriptor::BitmapImageDescriptor(CheckedRef<BitmapImageSource>&& source)
+    : m_source(WTFMove(source))
```
No `WTFMove()`

cdumez commented on Source/WebCore/platform/graphics/BitmapImageDescriptor.h:
```diff
@@ -123,7 +124,7 @@ class BitmapImageDescriptor {
     mutable SubsamplingLevel m_maximumSubsamplingLevel { SubsamplingLevel::Default };
 
-    BitmapImageSource& m_source;
+    CheckedRef<BitmapImageSource> m_source;
```
I think we should mark this as `const` since it can never be updated, it helps our static analyzer figure out what is safe:
```
const CheckedRef<BitmapImageSource> m_source;
```

iidmsa replied:
Done.

cdumez commented on Source/WebCore/platform/graphics/BitmapImageDescriptor.h:
```diff
@@ -42,7 +43,7 @@ class NativeImage;
 class BitmapImageDescriptor {
 public:
-    BitmapImageDescriptor(BitmapImageSource&);
+    BitmapImageDescriptor(CheckedRef<BitmapImageSource>&&);
```
We should revert his change.

iidmsa replied:
Done.

cdumez reviewed (approved):
LGTM assuming the bots are happy.

iidmsa committed:
Address Safer cpp failures in BitmapImageDescriptor
https://bugs.webkit.org/show_bug.cgi?id=288767
rdar://problem/145790133

Reviewed by Chris Dumez.

Fix a clang static analyzer warning in BitmapImageDescriptor.

* Source/WebCore/platform/graphics/BitmapImageDescriptor.cpp:
(WebCore::BitmapImageDescriptor::BitmapImageDescriptor):
(WebCore::BitmapImageDescriptor::imageMetadata const):
(WebCore::BitmapImageDescriptor::primaryImageFrameMetadata const):
(WebCore::BitmapImageDescriptor::primaryNativeImageMetadata const):
(WebCore::BitmapImageDescriptor::sourceSize const):
(WebCore::BitmapImageDescriptor::singlePixelSolidColor const):
(WebCore::BitmapImageDescriptor::maximumSubsamplingLevel const):
(WebCore::BitmapImageDescriptor::shouldUseQuickLookForFullscreen const):
(WebCore::BitmapImageDescriptor::isSpatial const):
* Source/WebCore/platform/graphics/BitmapImageDescriptor.h:

Canonical link: https://commits.webkit.org/291436@main
